File: backend/app/bootstrap/application.py
from __future__ import annotations

from dataclasses import dataclass
from collections.abc import Callable
from pathlib import Path

from app.config.settings import Settings
from app.infra.artifacts import FileArtifactStore
from app.infra.database import Database
from app.infra.persistence_root import PersistenceRoot
from app.lifecycle.service import ApplicationLifecycle
from app.lifecycle.worker_lease import WorkerLeaseStore
from app.observability.trace import TraceContext


@dataclass
class ApplicationRuntime:
    """汇总后续业务任务消费的基础运行时组件。"""
    settings: Settings
    database: Database
    lifecycle: ApplicationLifecycle
    leases: WorkerLeaseStore
    artifact_store: FileArtifactStore
    trace_context_factory: Callable[[], TraceContext]


def build_runtime(persistent_root: Path, *, test_mode: bool = False) -> ApplicationRuntime:
    """初始化持久化根目录并构造数据库、生命周期和证据存储组件。"""
    settings = Settings(persistent_root=persistent_root)
    root = PersistenceRoot(
        settings.persistent_root,
        app_version=settings.app_version,
        schema_revision=settings.current_schema_revision,
    )
    database = Database(
        settings.database_path,
        persistent_root=settings.persistent_root,
        app_version=settings.app_version,
        schema_revision=settings.current_schema_revision,
    )
    # 修改说明：Task 2 迁移前必须先检查真实 revision；只有新库/成功升级才更新 manifest。
    root.initialize_database(database)
    leases = WorkerLeaseStore(database)
    lifecycle = ApplicationLifecycle(database, leases)
    return ApplicationRuntime(
        settings,
        database,
        lifecycle,
        leases,
        FileArtifactStore(settings.artifact_path),
        TraceContext.new,
    )

File: backend/src/bootstrap/application.ts
import { Settings } from "../config/settings.js";
import { FileArtifactStore } from "../infra/artifacts.js";
import { Database } from "../infra/database.js";
import { PersistenceRoot } from "../infra/persistence-root.js";
import { ApplicationLifecycle } from "../lifecycle/service.js";
import { WorkerLeaseStore } from "../lifecycle/worker-lease.js";
import { TraceContext } from "../observability/trace.js";

/** 汇总后续业务任务消费的基础运行时组件。 */
export type ApplicationRuntime = {
  settings: Settings;
  database: Database;
  lifecycle: ApplicationLifecycle;
  leases: WorkerLeaseStore;
  artifactStore: FileArtifactStore;
  traceContextFactory: () => TraceContext;
};
/** 初始化持久化根目录并构造数据库、生命周期和证据存储组件。 */
export function buildRuntime(
  persistentRoot: string,
  options: { testMode?: boolean } = {},
): ApplicationRuntime {
  const settings = new Settings({ persistentRoot });
  const root = new PersistenceRoot(settings.persistentRoot, {
    appVersion: settings.appVersion,
    schemaRevision: settings.currentSchemaRevision,
  });
  const database = new Database(settings.databasePath, {
    persistentRoot: settings.persistentRoot,
    appVersion: settings.appVersion,
    schemaRevision: settings.currentSchemaRevision,
  });
  root.initializeDatabase(database);
  const leases = new WorkerLeaseStore(database);
  return {
    settings,
    database,
    lifecycle: new ApplicationLifecycle(database, leases),
    leases,
    artifactStore: new FileArtifactStore(
      settings.artifactPath,
      settings.artifactMaxSizeBytes,
    ),
    traceContextFactory: TraceContext.new,
  };
}
